### ISSUE TITLE
Better greedy layout computation for generators

### ISSUE THREAD

**Issue by tmandry** [rust-lang/rust#62575]
I've been thinking about a new approach for computing the layout of generators, since I merged #59897. Actually, I've been thinking about it before that, but the approach there felt a bit safer to start with.

The idea is that we use the struct layout computation, rather than the enum computation, as a starting point, and add in the fact that fields can overlap. The basic algorithm is simple:

- We iterate through every field, in descending order of alignment and number of conflicts with other fields*.
- For each field, pick the first offset (with respect to its alignment) that it's still allowed to occupy, given the other fields that we've already placed.

Because we're working in descending order of alignment (and alignments are all powers of 2), we should be able to use a "segment-tree-like" data structure with a bitset of fields that are still allowed to be placed at a given position. This means that updating the tree is logarithmic in the number of alignment levels, and so is searching for the first position a field is allowed to occupy. 

This greedy algorithm should be an improvement over the current algorithm, which doesn't attempt to overlap any fields that are live across more than one yield in the generator. It also wouldn't force "overlap-zone" fields to always come after "non-overlap" fields, which means we can do a better job of eliminating padding. Finally, the logic should be easier to follow in code than the current optimization (once the data structure is written).

This problem is NP-complete, so there might be better algorithms than this out there. I'd love to hear suggestions!

---

(*) The number of conflicts is the number of other fields in this struct that cannot be overlapped with the current field. The exact heuristic may change. Today, we use number of conflicts to help us decide which fields get to stay in an "overlap zone" and which get kicked out. Then, we use order of alignment to lay out the fields in each zone.

**Comment by Eulogizethesun** [comment]
When I create and use a generator, can I get the specific memory layout of the generator?
I want to get the current state of a generator in sometime . So the state information maybe stored in the specific memory.
I also want to know how to get the layout information of a Future function, becoause Future is created from generator.

**Comment by bjorn3** [comment]
No, you can't form rust code. If you are a debugger, you can read the debuginfo. Why do you want to get the current state of a generator?

**Comment by memoryruins** [comment]
> I also want to know how to get the layout information of a Future function

@Eulogizethesun

```rust
#![feature(rustc_attrs, type_alias_impl_trait)]

use core::future::Future;

#[rustc_layout(debug)]
type T = impl Future;

fn f() -> T {
    async {}
}
```

will emit a bit of layout info:

<details>


```
error: layout_of(std::future::from_generator::GenFuture<[static generator@src/lib.rs:9:11: 9:13 {}]>) = Layout {
    fields: Arbitrary {
        offsets: [
            Size {
                raw: 0,
            },
        ],
        memory_index: [
            0,
        ],
    },
    variants: Single {
        index: 0,
    },
    abi: Aggregate {
        sized: true,
    },
    largest_niche: Some(
        Niche {
            offset: Size {
                raw: 0,
            },
            scalar: Scalar {
                value: Int(
                    I8,
                    false,
                ),
                valid_range: 0..=2,
            },
        },
    ),
    align: AbiAndPrefAlign {
        abi: Align {
            pow2: 0,
        },
        pref: Align {
            pow2: 3,
        },
    },
    size: Size {
        raw: 1,
    },
}
```

</details>

**Comment by Eulogizethesun** [comment]
> No, you can't from rust code. If you are a debugger, you can read the debuginfo. Why do you want to get the current state of a generator?

Actually, I'm programm in assembly code. It makes sense to get the current state of a generator.Of cource I can use *size_of_val* to get the size of a Future function, but I dont' know what information is stored in which byte.
So the information is recorded in debuginfo, but how can I get it?

**Comment by jonas-schievink** [comment]
This discussion is unrelated to this issue, which is about improving the layout algorithm. Please more it somewhere else.
